br/pkg/restore/log_client/client.go

```diff
@@ -246,36 +250,43 @@ func (rc *LogClient) InitClients(ctx context.Context, backend *backuppb.StorageB
 	rc.fileImporter = NewLogFileImporter(metaClient, importCli, backend)
 }
 
-func (rc *LogClient) InitCheckpointMetadataForLogRestore(ctx context.Context, taskName string, gcRatio string) (string, error) {
+func (rc *LogClient) InitCheckpointMetadataForLogRestore(
+	ctx context.Context,
+	startTS, restoredTS uint64,
+	gcRatio string,
+	tiflashRecorder *tiflashrec.TiFlashRecorder,
+) (string, error) {
 	rc.useCheckpoint = true
 
-	// it shows that the user has modified gc-ratio, if `gcRatio` doesn't equal to "1.1".
-	// update the `gcRatio` for checkpoint metadata.
-	if gcRatio == utils.DefaultGcRatioVal {
-		// if the checkpoint metadata exists in the external storage, the restore is not
-		// for the first time.
-		exists, err := checkpoint.ExistsRestoreCheckpoint(ctx, rc.storage, taskName)
+	// if the checkpoint metadata exists in the external storage, the restore is not
+	// for the first time.
+	if checkpoint.ExistsLogRestoreCheckpointMetadata(ctx, rc.dom) {
+		// load the checkpoint since this is not the first time to restore
+		meta, err := checkpoint.LoadCheckpointMetadataForLogRestore(ctx, rc.se.GetSessionCtx().GetRestrictedSQLExecutor())
 		if err != nil {
 			return "", errors.Trace(err)
 		}
 
-		if exists {
-			// load the checkpoint since this is not the first time to restore
-			meta, err := checkpoint.LoadCheckpointMetadataForRestore(ctx, rc.storage, taskName)
-			if err != nil {
-				return "", errors.Trace(err)
-			}
-
-			log.Info("reuse gc ratio from checkpoint metadata", zap.String("gc-ratio", gcRatio))
-			return meta.GcRatio, nil
-		}
+		log.Info("reuse gc ratio from checkpoint metadata", zap.String("gc-ratio", gcRatio))
+		return meta.GcRatio, nil
 	}
 
 	// initialize the checkpoint metadata since it is the first time to restore.
-	log.Info("save gc ratio into checkpoint metadata", zap.String("gc-ratio", gcRatio))
-	if err := checkpoint.SaveCheckpointMetadataForRestore(ctx, rc.storage, &checkpoint.CheckpointMetadataForRestore{
-		GcRatio: gcRatio,
-	}, taskName); err != nil {
+	var items map[int64]model.TiFlashReplicaInfo
+	if tiflashRecorder != nil {
+		items = tiflashRecorder.GetItems()
+	}
+	log.Info("save gc ratio into checkpoint metadata",
+		zap.Uint64("start-ts", startTS), zap.Uint64("restored-ts", restoredTS), zap.Uint64("rewrite-ts", rc.currentTS),
+		zap.String("gc-ratio", gcRatio), zap.Int("tiflash-item-count", len(items)))
+	if err := checkpoint.SaveCheckpointMetadataForLogRestore(ctx, rc.se, &checkpoint.CheckpointMetadataForLogRestore{
```

Comment:
maybe just return this function directly?

Reply:
Just add trace to convenient to locate the error.